### ISSUE TITLE
Compiler loses type narrowing info within nested functions

### ISSUE THREAD

**Issue by trevorade** [google/closure-compiler#2193]
The following example:

```javascript
const /** number|undefined */ num = 2;
if (num) {
  const /** number */ correct = num;
  const f = () => {
    const /** number */ unexpectedWarning = num;
  };
}
```

Produces the following warning:
```
input0:5: WARNING - initializing variable
found   : (number|undefined)
required: number
    const /** number */ unexpectedWarning = num;
                                            ^^^

0 error(s), 1 warning(s), 100.0% typed
```

I can understand why that warning may happen for `var` or `let` (the value of the variable could change before the function is called) but it seems like it should be smarter about dealing with `const`.

This example with default parameters has the same issue:

```javascript
/** @param {number=} num */
function f(num = 2) {
  const /** number */ correct = num;
  const func = () => {
    const /** number */ unexpectedWarning = num;
  };
}
```

In this case, I suppose there isn't much that can be done as it would be legal to later write: `num = undefined;`.  Maybe the compiler could check for modifications to the parameter value?  Seems hard though.

In this case, if default parameter values are used, it seems like the parameter type should always be `number` rather than `number|undefined`.  i.e., this would be illegal:

```javascript
/** @param {number=} num */
function f(num = undefined) {}
```

You'd need to write:

```javascript
/** @param {(number|undefined)=} num */
function f(num = undefined) {}
```

**Comment by Victorystick** [comment]
I've also encountered this issue. Neat examples. The only problem I have with considering
```js
/** @param {number=} num */
function f(num = undefined) {}
```
illegal is that it could break code depending on the existing behavior. The compiler currently accepts it, including passing arguments of types `number` and/or `undefined`.
```js
f(1337); // Ok
f(undefined); // Also, ok
```
But I agree that it would be good to be able to tell what type a variable has by looking at its type signature. :stuck_out_tongue: So far, a trailing `=` has implied `|undefined`, but default values may invalidate that assumption.

If nothing else, I'd like the compiler to restrict the type depending on what types the variable is assigned. In the example above, `num` would have type `(number|undefined)` since it's declared to have type `number` with an `undefined` default value. In the example below however, `num` would strictly have type `number` since any default values passed will be replaced with `2`.
```js
/** @param {number=} num */
function f(num = 2) {}
```

**Comment by trevorade** [comment]
@Victorystick, 

> If nothing else, I'd like the compiler to restrict the type depending on what types the variable is assigned.

Yeah, that would be fine with me.  If the implicit `{|undefined}` was removed when the param default value is something other than `undefined`, the bug that brought this to my attention would be fixed.
